File: .tmp_c4000/batch_04_go/3478_choose_k_elements_with_maximum_sum.go
// LeetCode 3478 - Choose K Elements With Maximum Sum
// https://leetcode.com/problems/choose-k-elements-with-maximum-sum/

import "container/heap"
import "sort"

type mh3478 []int

func (h mh3478) Len() int            { return len(h) }
func (h mh3478) Less(i, j int) bool  { return h[i] < h[j] }
func (h mh3478) Swap(i, j int)       { h[i], h[j] = h[j], h[i] }
func (h *mh3478) Push(x interface{}) { *h = append(*h, x.(int)) }
func (h *mh3478) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[:n-1]
	return x
}

func findMaxSum(nums1 []int, nums2 []int, k int) []int64 {
	n := len(nums1)
	type item struct{ v1, v2, i int }
	arr := make([]item, n)
	for i := range nums1 {
		arr[i] = item{nums1[i], nums2[i], i}
	}
	sort.Slice(arr, func(i, j int) bool { return arr[i].v1 < arr[j].v1 })
	ans := make([]int64, n)
	h := &mh3478{}
	var sum int64
	j := 0
	for i := 0; i < n; {
		v := arr[i].v1
		start := i
		for i < n && arr[i].v1 == v {
			i++
		}
		// all with smaller nums1 already in heap
		for t := start; t < i; t++ {
			ans[arr[t].i] = sum
		}
		for t := start; t < i; t++ {
			heap.Push(h, arr[t].v2)
			sum += int64(arr[t].v2)
			if h.Len() > k {
				sum -= int64(heap.Pop(h).(int))
			}
		}
		_ = j
	}
	return ans
}

File: .tmp_py4013/batch_13_dump/3478_choose_k_elements_with_maximum_sum.js
// LeetCode 3478 - Choose K Elements With Maximum Sum
// https://leetcode.com/problems/choose-k-elements-with-maximum-sum/

var findMaxSum = function(nums1, nums2, k) {
    const n = nums1.length;
    const arr = [];
    for (let i = 0; i < n; i++) arr.push([nums1[i], nums2[i], i]);
    arr.sort((a, b) => a[0] - b[0]);
    const ans = new Array(n);
    const h = [];
    let sum = 0;
    const push = (v) => { h.push(v); h.sort((a, b) => a - b); };
    const poll = () => h.shift();
    for (let i = 0; i < n; ) {
        const v = arr[i][0];
        const start = i;
        while (i < n && arr[i][0] === v) i++;
        for (let t = start; t < i; t++) ans[arr[t][2]] = sum;
        for (let t = start; t < i; t++) {
            push(arr[t][1]);
            sum += arr[t][1];
            if (h.length > k) sum -= poll();
        }
    }
    return ans;
};
